cynthion/python/examples/tutorials/gateware-usb-device-04.py

```diff
@@ -0,0 +1,272 @@
+                    with m.If(interface.status_requested):
+                        m.d.comb += self.send_zlp()
+                        m.next = "IDLE"
+
+                with m.State("HANDLE_GET_USER_BUTTON"):
+                    # write the state of the user button into a local data register
+                    data = Signal(8)
+                    m.d.comb += data[0].eq(user_button)
+
+                    # transmit our data using a built-in handler function that
+                    # automatically advances the FSM back to the 'IDLE' state on
+                    # completion
+                    self.handle_simple_data_request(m, serializer, data)
+
+        return m
+
+
+class StreamingMemoryStore(Elaboratable):
+    """ A simple memory storage module that exposes a read/write streaming interface. """
+
+    def __init__(self, stream_out: StreamInterface, stream_in: StreamInterface):
+        self.stream_out = stream_out
+        self.stream_in  = stream_in
+
+        # high when a memory write is in process
+        self.write_active = Signal()
+
+    def elaborate(self, platform):
+        m = Module()
+
+        # create a memory we will use as a data source/sink for our bulk endpoints
+        m.submodules.ram = ram = Memory(
+            width = 8,
+            depth = MAX_PACKET_SIZE,
+            init  = [0] * MAX_PACKET_SIZE
+        )
+        w_port = ram.write_port(domain="usb")
+        r_port = ram.read_port(domain="usb")
+
+        # set the write_active status to the write port's enable status
+        m.d.comb += self.write_active.eq(w_port.en)
+
+        # shortcuts
+        stream_out  = self.stream_out
+        stream_in   = self.stream_in
+
+        # - EP 0x01 OUT logic  ------------------------------------------------
+
+        # let the stream know we're always ready to start reading
+        m.d.comb += stream_out.ready.eq(1)
+
+        # wire the payload from the host up to our memory write port
+        m.d.comb += w_port.data.eq(stream_out.payload)
+
+        # read each byte coming in on the stream and write it to memory
+        with m.If(stream_out.valid & stream_out.ready):
+            m.d.comb += w_port.en.eq(1)
+            m.d.usb += w_port.addr.eq(w_port.addr + 1);
+        with m.Else():
+            m.d.comb += w_port.en.eq(0)
+            m.d.usb += w_port.addr.eq(0)
+
+        # - EP 0x82 IN logic  -------------------------------------------------
+
+        # wire the payload to the host up to our memory read port
+        m.d.comb += stream_in.payload.eq(r_port.data)
+
+        # when the stream is ready and the write port is not active,
+        # read each byte from memory and write it out to the stream
+        with m.If(stream_in.ready & ~w_port.en):
+            m.d.usb += stream_in.valid.eq(1)
+            m.d.usb += r_port.addr.eq(r_port.addr + 1)
+        with m.Else():
+            m.d.usb += stream_in.valid.eq(0)
+            m.d.usb += r_port.addr.eq(0)
```

Comment:
I suggest a simpler example here with a `SyncFIFO`. Note it also blocks the stream when the FIFO is full:

```python
m.submodules.mem = mem = DomainRenamer("usb")(fifo.SyncFIFO(width=8, depth=MAX_PACKET_SIZE))

...

m.d.comb += mem.w_data.eq(stream_out.payload)
m.d.comb += mem.w_en.eq(stream_out.valid)
m.d.comb += stream_out.ready.eq(mem.w_rdy)

m.d.comb += stream_in.payload.eq(mem.r_data)
m.d.comb += stream_in.valid.eq(mem.r_rdy)
m.d.comb += mem.r_en.eq(stream_in.ready)
```
Since Amaranth HDL v0.5.0, you can do this instead:

```python
wiring.connect(m, stream_out, mem.w_stream)
wiring.connect(m, mem.r_stream, stream_in)
```

Reply:
https://github.com/greatscottgadgets/cynthion/pull/197/commits/2dd24e1c93af9048ce4181d954cfe6b4f2c6880a